File: blueprints/api/assets.py
"""
Complete Assets API Blueprint - Full replacement to fix schema issues
Aligned with simplified BusinessAsset model (no current_value field)
"""

from flask import Blueprint, request, jsonify
from datetime import datetime
from models import db, BusinessAsset, BusinessTransaction

# Create blueprint
assets_api_bp = Blueprint('assets_api', __name__, url_prefix='/api/assets')

@assets_api_bp.route('', methods=['POST'])
def add_asset():
    """Add new business asset with automatic expense transaction using selected expense category"""
    
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'success': False, 'error': 'No data provided'}), 400
        
        # Validate required fields
        required_fields = ['name', 'asset_type', 'purchase_date', 'purchase_price', 'expense_category']
        missing_fields = [field for field in required_fields if not data.get(field)]
        
        if missing_fields:
            return jsonify({
                'success': False, 
                'error': f'Missing required fields: {", ".join(missing_fields)}'
            }), 400
        
        # Parse and validate purchase date
        try:
            purchase_date = datetime.strptime(data['purchase_date'], '%Y-%m-%d').date()
        except ValueError:
            return jsonify({
                'success': False, 
                'error': 'Invalid purchase date format. Use YYYY-MM-DD'
            }), 400
        
        # Parse and validate purchase price
        try:
            purchase_price = float(data['purchase_price'])
            if purchase_price <= 0:
                return jsonify({
                    'success': False, 
                    'error': 'Purchase price must be greater than 0'
                }), 400
        except (ValueError, TypeError):
            return jsonify({
                'success': False, 
                'error': 'Invalid purchase price format'
            }), 400
        
        # Create new asset with simplified schema (no current_value field)
        asset = BusinessAsset(
            name=data['name'].strip(),
            description=data.get('description', '').strip(),
            asset_category=data.get('expense_category', '').strip(),  # Now uses expense category
            asset_type=data['asset_type'].strip(),
            purchase_date=purchase_date,
            purchase_price=purchase_price,
            is_active=data.get('is_active', True)  # Default to active
        )
        
        db.session.add(asset)
        db.session.flush()  # Get the asset ID
        
        # Create automatic expense transaction using selected categories
        expense_transaction = BusinessTransaction(
            date=purchase_date,
            description=f"Asset Purchase: {asset.name}",
            amount=purchase_price,
            category=data['expense_category'],  # Use selected expense category
            sub_category=data.get('expense_sub_category', 'Asset Purchase'),  # Use selected sub-category or default
            transaction_type="Expense",
            account_name=data.get('account_name', 'Business Checking'),  # Use selected account
            notes=f"Automatic transaction for asset purchase (Asset ID: {asset.id})"
        )
        
        db.session.add(expense_transaction)
        db.session.commit()
        
        print(f"✅ Asset added successfully: {asset.name} (ID: {asset.id})")
        print(f"✅ Expense transaction created: ${purchase_price} in category '{data['expense_category']}' (ID: {expense_transaction.id})")
        
        return jsonify({
            'success': True,
            'message': 'Asset added successfully with expense transaction!',
            'asset': asset.to_dict(),
            'transaction': expense_transaction.to_dict(),
            'asset_id': asset.id,
            'transaction_id': expense_transaction.id
        })
        
    except Exception as e:
        print(f"❌ Error adding asset: {e}")
        import traceback
        traceback.print_exc()
        db.session.rollback()
        return jsonify({'success': False, 'error': str(e)}), 500

@assets_api_bp.route('/<int:asset_id>', methods=['PUT'])
def update_asset(asset_id):
    """Update existing asset"""
    
    try:
        asset = BusinessAsset.query.get(asset_id)
        
        if not asset:
            return jsonify({'success': False, 'error': 'Asset not found'}), 404
        
        data = request.get_json()
        
        if not data:
            return jsonify({'success': False, 'error': 'No data provided'}), 400
        
        # Update basic fields if provided
        if 'name' in data:
            asset.name = data['name'].strip()
        if 'description' in data:
            asset.description = data['description'].strip()
        if 'expense_category' in data:
            asset.asset_category = data['expense_category'].strip()  # Store expense category as asset category
        if 'asset_type' in data:
            asset.asset_type = data['asset_type'].strip()
        if 'is_active' in data:
            asset.is_active = bool(data['is_active'])
        
        # Update purchase date if provided
        if 'purchase_date' in data and data['purchase_date']:
            try:
                asset.purchase_date = datetime.strptime(data['purchase_date'], '%Y-%m-%d').date()
            except ValueError:
                return jsonify({
                    'success': False, 
                    'error': 'Invalid purchase date format. Use YYYY-MM-DD'
                }), 400
        
        # Update purchase price if provided
        if 'purchase_price' in data:
            try:
                purchase_price = float(data['purchase_price'])
                if purchase_price <= 0:
                    return jsonify({
                        'success': False, 
                        'error': 'Purchase price must be greater than 0'
                    }), 400
                asset.purchase_price = purchase_price
            except (ValueError, TypeError):
                return jsonify({
                    'success': False, 
                    'error': 'Invalid purchase price format'
                }), 400
        
        db.session.commit()
        
        print(f"✅ Asset updated successfully: {asset.name} (ID: {asset.id})")
        
        return jsonify({
            'success': True,
            'message': 'Asset updated successfully!',
            'asset': asset.to_dict()
        })
        
    except Exception as e:
        print(f"❌ Error updating asset: {e}")
        db.session.rollback()
        return jsonify({'success': False, 'error': str(e)}), 500

@assets_api_bp.route('/<int:asset_id>', methods=['DELETE'])
def delete_asset(asset_id):
    """Delete an asset"""
    
    try:
        asset = BusinessAsset.query.get(asset_id)
        
        if not asset:
            return jsonify({'success': False, 'error': 'Asset not found'}), 404
        
        asset_name = asset.name
        db.session.delete(asset)
        db.session.commit()
        
        print(f"✅ Asset deleted successfully: {asset_name} (ID: {asset_id})")
        
        return jsonify({
            'success': True,
            'message': f'Asset "{asset_name}" deleted successfully!'
        })
        
    except Exception as e:
        print(f"❌ Error deleting asset: {e}")
        db.session.rollback()
        return jsonify({'success': False, 'error': str(e)}), 500

@assets_api_bp.route('', methods=['GET'])
def get_assets():
    """Get all assets with optional filtering"""
    
    try:
        # Get query parameters
        active_only = request.args.get('active_only', 'true').lower() == 'true'
        category = request.args.get('category')
        
        # Build query
        query = BusinessAsset.query
        
        if active_only:
            query = query.filter_by(is_active=True)
        
        if category:
            query = query.filter_by(asset_category=category)
        
        # Get assets
        assets = query.order_by(BusinessAsset.purchase_date.desc()).all()
        
        # Convert to dictionaries
        assets_data = [asset.to_dict() for asset in assets]
        
        return jsonify({
            'success': True,
            'assets': assets_data,
            'count': len(assets_data)
        })
        
    except Exception as e:
        print(f"❌ Error getting assets: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

@assets_api_bp.route('/<int:asset_id>', methods=['GET'])
def get_asset(asset_id):
    """Get single asset details"""
    
    try:
        asset = BusinessAsset.query.get(asset_id)
        
        if not asset:
            return jsonify({'success': False, 'error': 'Asset not found'}), 404
        
        return jsonify({
            'success': True,
            'asset': asset.to_dict()
        })
        
    except Exception as e:
        print(f"❌ Error getting asset: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

File: static/js/assets.js
/**
 * Complete Assets JavaScript - Full replacement to fix all issues
 * Aligned with simplified asset schema and expense category system
 */

/**
 * Enhanced function to save asset (replaces existing saveAsset function completely)
 */
function saveAsset() {
    console.log('💾 Saving asset with expense categories...');
    
    const form = document.getElementById('assetForm');
    if (!form || !form.checkValidity()) {
        form.reportValidity();
        return;
    }
    
    // Get all form values (only fields that exist in simplified schema)
    const assetData = {
        name: document.getElementById('assetName').value.trim(),
        asset_type: document.getElementById('assetType').value.trim(),
        purchase_date: document.getElementById('assetPurchaseDate').value,
        purchase_price: parseFloat(document.getElementById('assetPurchasePrice').value),
        expense_category: document.getElementById('assetExpenseCategory').value,
        expense_sub_category: document.getElementById('assetExpenseSubCategory').value,
        account_name: document.getElementById('assetAccountName').value,
        description: document.getElementById('assetDescription').value.trim(),
        is_active: document.getElementById('assetStatus').value === '1'
    };
    
    // Validate required fields
    if (!assetData.name || !assetData.asset_type || !assetData.purchase_date || 
        !assetData.purchase_price || !assetData.expense_category) {
        showAlert('Please fill in all required fields', 'danger');
        return;
    }
    
    // Check if editing existing asset
    const assetId = document.getElementById('assetId').value;
    const isEditing = assetId && assetId.trim() !== '';
    
    // Prepare API call
    const url = isEditing ? `/api/assets/${assetId}` : '/api/assets';
    const method = isEditing ? 'PUT' : 'POST';
    
    // Show loading state
    const saveButton = document.getElementById('saveAssetText');
    const originalText = saveButton.textContent;
    saveButton.textContent = 'Saving...';
    
    fetch(url, {
        method: method,
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(assetData)
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            console.log('✅ Asset saved successfully:', data);
            
            // Show success message
            showAlert(data.message || `Asset ${isEditing ? 'updated' : 'added'} successfully!`, 'success');
            
            // Close modal
            const modal = bootstrap.Modal.getInstance(document.getElementById('assetModal'));
            if (modal) modal.hide();
            
            // Reload assets table or page
            if (typeof loadAssets === 'function') {
                loadAssets();
            } else {
                // Fallback: reload page if loadAssets function doesn't exist
                setTimeout(() => window.location.reload(), 1000);
            }
            
            // Reset form
            form.reset();
            document.getElementById('assetId').value = '';
            
        } else {
            console.error('❌ Error saving asset:', data.error);
            showAlert(data.error || 'Error saving asset', 'danger');
        }
    })
    .catch(error => {
        console.error('❌ Network error saving asset:', error);
        showAlert('Network error. Please try again.', 'danger');
    })
    .finally(() => {
        // Restore button text
        saveButton.textContent = originalText;
    });
}

/**
 * Populate asset sub-categories based on expense category selection
 */
function populateAssetSubCategories() {
    const categorySelect = document.getElementById('assetExpenseCategory');
    const subCategorySelect = document.getElementById('assetExpenseSubCategory');
    
    if (!categorySelect || !subCategorySelect) return;
    
    const selectedCategory = categorySelect.value;
    
    // Sub-category mapping (matches the expense modal)
    const subCategoryMap = {
        "Cost of Goods Sold": [
            "Inventory Purchase",
            "Raw Materials",
            "Direct Labor",
            "Manufacturing Costs"
        ],
        "Marketing & Advertising": [
            "Online Advertising",
            "Print Advertising", 
            "Social Media Marketing",
            "Website & SEO",
            "Business Cards",
            "Promotional Materials"
        ],
        "Operations": [
            "Rent & Utilities",
            "Office Supplies",
            "Venue Rental",
            "Storage Costs",
            "Maintenance & Repairs"
        ],
        "Equipment & Supplies": [
            "Computer Equipment",
            "Office Furniture", 
            "POS Systems",
            "Software Licenses",
            "Tools & Equipment",
            "Asset Purchase"
        ],
        "Professional Services": [
            "Legal Services",
            "Accounting Services", 
            "Consulting",
            "Banking Fees",
            "Insurance"
        ],
        "Travel & Transport": [
            "Business Travel",
            "Vehicle Expenses",
            "Shipping & Delivery",
            "Transportation"
        ],
        "Other Expenses": [
            "Miscellaneous",
            "One-time Costs",
            "Unexpected Expenses"
        ]
    };
    
    // Clear existing sub-categories
    subCategorySelect.innerHTML = '<option value="">Select sub-category...</option>';
    
    if (selectedCategory && subCategoryMap[selectedCategory]) {
        const subCategories = subCategoryMap[selectedCategory];
        
        subCategories.forEach(subCategory => {
            const option = document.createElement('option');
            option.value = subCategory;
            option.textContent = subCategory;
            subCategorySelect.appendChild(option);
        });
        
        // Auto-select "Asset Purchase" if available for Equipment & Supplies
        if (selectedCategory === "Equipment & Supplies") {
            subCategorySelect.value = "Asset Purchase";
        }
    }
}

/**
 * Show the add asset modal with proper defaults
 */
function showAddAssetModal() {
    console.log('➕ Opening enhanced add asset modal...');
    
    // Reset form
    const form = document.getElementById('assetForm');
    if (form) form.reset();
    
    // Clear hidden ID field
    const assetIdInput = document.getElementById('assetId');
    if (assetIdInput) assetIdInput.value = '';
    
    // Set default values
    setAssetFormDefaults();
    
    // Update modal title and button
    const modalLabel = document.getElementById('assetModalLabel');
    const saveButton = document.getElementById('saveAssetText');
    
    if (modalLabel) modalLabel.textContent = 'Add New Asset';
    if (saveButton) saveButton.textContent = 'Save Asset';
    
    // Show modal
    const modal = document.getElementById('assetModal');
    if (modal && typeof bootstrap !== 'undefined') {
        const bsModal = new bootstrap.Modal(modal);
        bsModal.show();
        
        // Focus on asset name field after modal opens
        modal.addEventListener('shown.bs.modal', function() {
            const nameInput = document.getElementById('assetName');
            if (nameInput) {
                setTimeout(() => nameInput.focus(), 100);
            }
        }, { once: true });
    }
}

/**
 * Set default form values
 */
function setAssetFormDefaults() {
    // Set today's date as default
    const dateInput = document.getElementById('assetPurchaseDate');
    if (dateInput) {
        const today = new Date().toISOString().split('T')[0];
        dateInput.value = today;
    }
    
    // Set default account name
    const accountSelect = document.getElementById('assetAccountName');
    if (accountSelect) {
        accountSelect.value = 'Business Checking';
    }
    
    // Set default status to active
    const statusField = document.getElementById('assetStatus');
    if (statusField) {
        statusField.value = '1';
    }
    
    // Set default expense category to Equipment & Supplies
    const categorySelect = document.getElementById('assetExpenseCategory');
    if (categorySelect) {
        categorySelect.value = 'Equipment & Supplies';
        // Trigger sub-category population
        populateAssetSubCategories();
    }
}

/**
 * Edit an asset (fetch data and populate form)
 */
function editAsset(assetId) {
    console.log(`✏️ Editing asset ${assetId}...`);
    
    // Fetch asset data first
    fetch(`/api/assets/${assetId}`)
    .then(response => response.json())
    .then(data => {
        if (data.success && data.asset) {
            const asset = data.asset;
            
            // Populate form with asset data
            document.getElementById('assetId').value = asset.id;
            document.getElementById('assetName').value = asset.name || '';
            document.getElementById('assetType').value = asset.asset_type || '';
            document.getElementById('assetPurchaseDate').value = asset.purchase_date || '';
            document.getElementById('assetPurchasePrice').value = asset.purchase_price || '';
            document.getElementById('assetExpenseCategory').value = asset.asset_category || '';
            document.getElementById('assetAccountName').value = 'Business Checking'; // Default since not stored
            document.getElementById('assetDescription').value = asset.description || '';
            document.getElementById('assetStatus').value = asset.is_active ? '1' : '0';
            
            // Populate sub-categories based on category
            populateAssetSubCategories();
            
            // Update modal title and button
            const modalLabel = document.getElementById('assetModalLabel');
            const saveButton = document.getElementById('saveAssetText');
            
            if (modalLabel) modalLabel.textContent = 'Edit Asset';
            if (saveButton) saveButton.textContent = 'Update Asset';
            
            // Show modal
            const modal = document.getElementById('assetModal');
            if (modal && typeof bootstrap !== 'undefined') {
                const bsModal = new bootstrap.Modal(modal);
                bsModal.show();
            }
            
        } else {
            console.error('❌ Error fetching asset data:', data.error);
            showAlert('Error loading asset data', 'danger');
        }
    })
    .catch(error => {
        console.error('❌ Network error fetching asset:', error);
        showAlert('Network error loading asset', 'danger');
    });
}

/**
 * Delete an asset with confirmation
 */
function deleteAsset(assetId) {
    if (!confirm('Are you sure you want to delete this asset? This action cannot be undone.')) {
        return;
    }
    
    console.log(`🗑️ Deleting asset ${assetId}...`);
    
    fetch(`/api/assets/${assetId}`, {
        method: 'DELETE'
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            console.log('✅ Asset deleted successfully');
            showAlert(data.message || 'Asset deleted successfully', 'success');
            
            // Reload assets table
            if (typeof loadAssets === 'function') {
                loadAssets();
            } else {
                // Fallback: reload page
                setTimeout(() => window.location.reload(), 1000);
            }
        } else {
            console.error('❌ Error deleting asset:', data.error);
            showAlert(data.error || 'Error deleting asset', 'danger');
        }
    })
    .catch(error => {
        console.error('❌ Network error deleting asset:', error);
        showAlert('Network error. Please try again.', 'danger');
    });
}

/**
 * Show asset details in modal (simplified for new schema)
 */
function showAssetDetails(assetId) {
    console.log(`👁️ Showing details for asset ${assetId}...`);
    
    fetch(`/api/assets/${assetId}`)
    .then(response => response.json())
    .then(data => {
        if (data.success && data.asset) {
            const asset = data.asset;
            
            // Build asset details HTML (simplified for new schema)
            const detailsHtml = `
                <div class="asset-details">
                    <div class="row mb-3">
                        <div class="col-sm-4"><strong>Name:</strong></div>
                        <div class="col-sm-8">${asset.name}</div>
                    </div>
                    <div class="row mb-3">
                        <div class="col-sm-4"><strong>Type:</strong></div>
                        <div class="col-sm-8">${asset.asset_type}</div>
                    </div>
                    <div class="row mb-3">
                        <div class="col-sm-4"><strong>Category:</strong></div>
                        <div class="col-sm-8">${asset.asset_category}</div>
                    </div>
                    <div class="row mb-3">
                        <div class="col-sm-4"><strong>Purchase Date:</strong></div>
                        <div class="col-sm-8">${new Date(asset.purchase_date).toLocaleDateString()}</div>
                    </div>
                    <div class="row mb-3">
                        <div class="col-sm-4"><strong>Purchase Price:</strong></div>
                        <div class="col-sm-8">$${parseFloat(asset.purchase_price).toFixed(2)}</div>
                    </div>
                    <div class="row mb-3">
                        <div class="col-sm-4"><strong>Status:</strong></div>
                        <div class="col-sm-8">
                            <span class="badge ${asset.is_active ? 'bg-success' : 'bg-secondary'}">
                                ${asset.is_active ? 'Active' : 'Disposed'}
                            </span>
                        </div>
                    </div>
                    ${asset.description ? `
                    <div class="row mb-3">
                        <div class="col-sm-4"><strong>Description:</strong></div>
                        <div class="col-sm-8">${asset.description}</div>
                    </div>
                    ` : ''}
                </div>
            `;
            
            // Show in modal
            document.getElementById('assetDetailsContent').innerHTML = detailsHtml;
            
            const modal = document.getElementById('assetDetailsModal');
            if (modal && typeof bootstrap !== 'undefined') {
                const bsModal = new bootstrap.Modal(modal);
                bsModal.show();
            }
            
        } else {
            console.error('❌ Error fetching asset details:', data.error);
            showAlert('Error loading asset details', 'danger');
        }
    })
    .catch(error => {
        console.error('❌ Network error fetching asset details:', error);
        showAlert('Network error loading asset details', 'danger');
    });
}

/**
 * Utility function to show alerts
 */
function showAlert(message, type = 'info') {
    // Create alert element
    const alertDiv = document.createElement('div');
    alertDiv.className = `alert alert-${type} alert-dismissible fade show`;
    alertDiv.setAttribute('role', 'alert');
    alertDiv.innerHTML = `
        ${message}
        <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
    `;
    
    // Find a good place to show the alert
    let container = document.querySelector('.alert-container');
    if (!container) {
        container = document.querySelector('.container-fluid') || document.body;
    }
    
    // Insert at the top
    container.insertBefore(alertDiv, container.firstChild);
    
    // Auto-dismiss after 5 seconds
    setTimeout(() => {
        if (alertDiv.parentNode) {
            alertDiv.remove();
        }
    }, 5000);
}

// Initialize on page load
document.addEventListener('DOMContentLoaded', function() {
    console.log('🏢 Assets JavaScript loaded and ready');
});